pkg/apis/migration/migration.go

```diff
@@ -0,0 +1,351 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	migration "sigs.k8s.io/vsphere-csi-driver/pkg/apis/migration/v1alpha1"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/vsphere"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/common"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+	k8s "sigs.k8s.io/vsphere-csi-driver/pkg/kubernetes"
+)
+
+const (
+	timeout  = 60 * time.Second
+	pollTime = 5 * time.Second
+)
+
+// ErrVolumeNotRegistered is returned when requested volumepath is not found from cache and K8S CR List
+var ErrVolumeNotRegistered = errors.New("volume is not registered to CNS")
+
+// VolumeMigrationService exposes interfaces to manage VolumeInfo cache and CR
+// for VolumePath, VolumeID and VolumeName mapping
+type VolumeMigrationService interface {
+	// GetVolumeInfo returns CnsvSphereVolumeMigrationSpec for given VolumePath from local cache
+	// if CnsvSphereVolumeMigrationSpec is not available in the cache, K8s CR look up will be performed, if that too
+	// can not retrieve VolumeInfo, ErrVolumeNotRegistered will be returned
+	GetVolumeInfo(ctx context.Context, volumePath string) (migration.CnsvSphereVolumeMigrationSpec, error)
+
+	// SaveVolumeInfo helps create CR for given cnsvSphereVolumeMigration
+	// this func also update local cache with supplied cnsvSphereVolumeMigration, after successful creation of CR
+	SaveVolumeInfo(ctx context.Context, cnsvSphereVolumeMigration migration.CnsvSphereVolumeMigration) error
+
+	// DeleteVolumeInfo helps delete cached mapping of volumePath to VolumeInfo and
+	// also helps delete K8s Custom Resource for specified volumePath
+	DeleteVolumeInfo(ctx context.Context, volumePath string) error
+
+	// LoadAllVolumeInfo helps cache all CR for CnsvSphereVolumeMigration
+	LoadAllVolumeInfo(ctx context.Context) error
+
+	// Register Volume takes in-tree VolumePath and helps register Volume with CNS
+	RegisterVolume(ctx context.Context, volumePath string, manager *common.Manager) (migration.CnsvSphereVolumeMigration, error)
+}
+
+// VolumeMigration holds migrated volume information and provides functionality around it.
+type volumeMigration struct {
+	// Volume Path to volumeInfo map
+	volumePathToVolumeInfo sync.Map
+	// migrationClient helps operate on CnsvSphereVolumeMigration custom resource
+	migrationClient client.Client
+}
+
+// onceForInTreeVolumeMigrationService is used for initializing the VolumeMigrationService singleton.
+var onceForVolumeMigrationService sync.Once
+
+// volumeMigrationInstance is instance of volumeMigration and implements interface for VolumeMigrationService
+var volumeMigrationInstance *volumeMigration
+
+// GetVolumeMigrationService returns the singleton VolumeMigrationService
+func GetVolumeMigrationService(ctx context.Context) (VolumeMigrationService, error) {
+	var err error
+	log := logger.GetLogger(ctx)
+	onceForVolumeMigrationService.Do(func() {
```

Comment:
`sync.Once()` is to ensure that service loads CRs and cache only once. 

At present only controller and syncer are using this service. both are in two different container, so `sync.Once()` is not to protect call across containers. 

I kept this in case if at all we need an instance of migration service and we call GetVolumeMigrationService again in the same container, we don't reinitialize cache and perform initializations again.